proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,477 @@
+    otherwise throw `InvalidArgument`
+  - Change the slot in the program account to the current slot
+  - Change the status stored in the program account to `NeverBeenDeployed`
+  - Copy the new authority address into the program account
+
+#### SetAuthority
+
+- Instruction accounts:
+  - `[writable]` The program account to change the authority of.
+  - `[signer]` The current authority of the program.
+  - `[signer]` The new authority of the program.
+- Instruction data:
+  - Enum variant `1u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least three instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header
+  - Check that the new authority (instruction account at index 2) has signed,
+    otherwise throw `MissingRequiredSignature`
+  - Check that the current authority is different from the new authority,
+    otherwise throw `InvalidArgument`
+  - Copy the new authority address into the program account
+
+#### Finalize
+
+- Instruction accounts:
+  - `[writable]` The program account to finalize.
+  - `[signer]` The current authority of the program.
+- Instruction data:
+  - Enum variant `2u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header
+  - Check that the status stored in the program account is `Deployed` or
+    that the status is `Retracted`,
+    otherwise throw `InvalidArgument`
+  - Change the status stored in the program account to `Finalized`
+
+#### SetAccountLength
+
+- Instruction accounts:
+  - `[writable]` The program account to change the size of.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `3u32`
+  - `u32` The new size after the operation.
+- Behavior:
+  - Charge 32 + new_size_in_bytes / cpi_bytes_per_unit CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header,
+  but skip the signer check if the new size is greater than the current size
+  - Check that the status stored in the program account is
+  `NeverBeenDeployed` or `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Check that the new size in bytes is at least the header length (48 bytes),
+  otherwise throw `InvalidArgument`
+  - Set the length of the program account to the requested new size.
+  - Note: In CPI the maximum growth is limited to 10 KiB in ABI v1 and
+  0 bytes in ABI v0. Thus, this should be called from a top-level instruction.
+  For this purpose this operation does not check the signer when growing.
+
+#### Write
+
+- Instruction accounts:
+  - `[writable]` The program account to write to.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `4u32`
+  - `u32` Byte offset at which to write the given bytes
+  - `[u8]` Chunk of the programs executable file
```

Comment:
This looks different from loader-v3 which serializes the length of the chunk as a `u64` before the chunk data. Are you implying that the length isn't serialized here and that `chunk_length_in_bytes` is inferred from `ix.data.len - 4 bytes (enum variant) - 4 bytes (offset)`? I think we should match loader-v3 behavior and serialize the chunk length as well. But consider serializing the length as u16 or u32 instead of u64 to save some bytes

Reply:
What is the advantage of doing so? It would introduce additional failure cases of the explicit field and implicit calculation disagreeing.